depthai_ros_driver/include/depthai_ros_driver/device_ros.hpp

```diff
@@ -73,40 +137,57 @@ class DeviceROS : public dai::DeviceBase {
 
 protected:
     // create stream based on name at the time of subscription
-    template <class MsgType>
-    auto generate_cb_lambda(std::string name) {
-        auto conn = this->getConnection();
-        const auto core_sub_lambda = [&, stream = dai::XLinkStream{*conn, name, dai::XLINK_USB_BUFFER_MAX_SIZE}](
-                                             MsgType& msg, dai::RawBuffer& data) {
+    template <class MsgType, dai::DatatypeEnum DataType>
+    auto generate_cb_lambda(std::unique_ptr<dai::XLinkStream>& stream,
+                            msgpack::sbuffer& sbuf)
+            -> boost::function<void(const boost::shared_ptr<MsgType const>&)> {
+        const auto core_sub_lambda = [&stream, &sbuf](const boost::shared_ptr<MsgType const>& msg) {
             Guard guard([] { ROS_ERROR("Communication failed: Device error or misconfiguration."); });
 
-            // TODO:  convert msg to data here somehow
-            std::vector<std::uint8_t> serialized = dai::StreamPacketParser::serializeMessage(data);
-            stream.write(serialized);
+            // convert msg to data
+            msgpack::pack(sbuf, *msg);
+            PacketWriter writer(*stream);
+            writer.write(msg->data.data(), msg->data.size(), reinterpret_cast<uint8_t*>(sbuf.data()), sbuf.size(),
+                    static_cast<uint32_t>(DataType));
+
+            sbuf.clear();  // Prevent the sbuf data is accumeted
+
             guard.disable();
         };
+
         return core_sub_lambda;
     }
 
     template <class MsgType>
-    auto generate_pub_lambda(ros::NodeHandle& nh, std::string name, std::size_t q_size) {
-        auto conn = this->getConnection();
-        const auto pub_lambda = [this, pub = nh.advertise<MsgType>(name, q_size),
-                                        // no writing happens, so 1 is sufficient
-                                        stream = dai::XLinkStream{*conn, name, 1}]() {
+    auto generate_pub_lambda(ros::NodeHandle& nh, std::string name, std::size_t q_size) {  // name: copied
+        ros::Publisher pub = nh.advertise<MsgType>(name, q_size);
+
+        const auto pub_lambda = [this, pub, name]() {
+            auto conn = this->getConnection();
+            auto stream = dai::XLinkStream(*conn, name, 1);  // no writing happens, so 1 is sufficient
             Guard guard([] { ROS_ERROR("Communication failed: Device error or misconfiguration."); });
 
             while (this->_running) {
                 // block till data is read
                 PacketReader reader{stream};
-                auto data = reader.getData();
-                // @TODO convert data to ROS message type here, somehow
+                auto packet = reader.getPacket();
+
+                auto [ser_start, ser_size] = reader.serialRange();
+                msgpack::object_handle oh =
+                        msgpack::unpack(reinterpret_cast<const char*>(packet->data + ser_start), ser_size);
```

Comment:
Can we encapsulate this in the PacketReader API?

Eg: reader.msgpack_begin_ptr() and reader.msgpack_size() as const member functions?